controllers/route/route_certificaterequest_controller.go

```diff
@@ -0,0 +1,362 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type CertificateRequestReconciler struct {
+	outils.ReconcilerBase
+	Log            logr.Logger
+	controllerName string
+}
+
+func (c *CertificateRequestReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c.controllerName = "route_certificaterequest_controller"
+
+	isCertManagerRoute := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			route, ok := e.Object.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(route)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			route, ok := e.Object.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(route)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			newRoute, ok := e.ObjectNew.DeepCopyObject().(*routev1.Route)
+			if !ok {
+				return false
+			}
+			return hasCertManagerAnnotations(newRoute)
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	hasCRStatusChanged := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, ok := e.Object.(*cmapi.CertificateRequest)
+			return ok
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			crOld, ok := e.ObjectOld.(*cmapi.CertificateRequest)
+			if !ok {
+				return false
+			}
+			crNew, ok := e.ObjectNew.(*cmapi.CertificateRequest)
+			if !ok {
+				return false
+			}
+			oldConditions := make(map[cmapi.CertificateRequestConditionType]cmapi.CertificateRequestCondition)
+			for _, oldCon := range crOld.Status.Conditions {
+				oldConditions[oldCon.Type] = oldCon
+			}
+			for _, newCon := range crNew.Status.Conditions {
+				oldCon, found := oldConditions[newCon.Type]
+				if !found {
+					return true
+				}
+				if oldCon.Status != newCon.Status {
+					return true
+				}
+			}
+			return !bytes.Equal(crOld.Status.Certificate, crNew.Status.Certificate)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			_, ok := e.Object.(*cmapi.CertificateRequest)
+			return ok
+		},
+	}
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&routev1.Route{
+			TypeMeta: metav1.TypeMeta{Kind: "Route"},
+		}, builder.WithPredicates(isCertManagerRoute)).
+		// We will create cert-manager CertificateRequests to populate TLS routes, so
+		// we need to respond to the creation / update events them too.
+		Watches(&source.Kind{
+			Type: &cmapi.CertificateRequest{
+				TypeMeta: metav1.TypeMeta{Kind: "CertificateRequest"},
+			},
+		}, &handler.EnqueueRequestForOwner{
+			OwnerType: &routev1.Route{},
+		}, builder.WithPredicates(hasCRStatusChanged)).
```

Comment:
CertificateRequest spec and annotations are immutable enforced by the webhook. An update will only ever be a status change.This isn't needed.